service.go

```diff
@@ -0,0 +1,541 @@
+func (s *service) push(ctx context.Context, rec thread.Record, id thread.ID, lid peer.ID, settings *tserv.AddSettings) error {
+	var addrs []ma.Multiaddr
+	// Collect known writers
+	for _, l := range s.threads.ThreadInfo(settings.Thread).Logs {
+		if l.String() == lid.String() {
+			continue
+		}
+		addrs = append(addrs, s.threads.Addrs(settings.Thread, l)...)
+	}
+
+	// Add additional addresses
+	addrs = append(addrs, settings.Addrs...)
+
+	// Serialize and sign the record for transport
+	pbrec, err := cbor.RecordToProto(ctx, s.threads.dagService, rec)
+	if err != nil {
+		return err
+	}
+	payload, err := pbrec.Marshal()
+	if err != nil {
+		return err
+	}
+	sk := s.threads.getPrivKey()
+	if sk == nil {
+		return fmt.Errorf("could not find key for host")
+	}
+	sig, err := sk.Sign(payload)
+	if err != nil {
+		return err
+	}
+
+	var keyLog *pb.ProtoPeerID
+	logKey := s.threads.ReadKey(settings.Thread, settings.KeyLog)
+	if logKey != nil {
+		keyLog = &pb.ProtoPeerID{ID: settings.KeyLog}
+	}
+	req := &pb.PushRequest{
+		Header: &pb.PushRequest_Header{
+			From:         &pb.ProtoPeerID{ID: s.threads.host.ID()},
+			Signature:    sig,
+			Key:          &pb.ProtoPubKey{PubKey: sk.GetPublic()},
+			FollowKey:    s.threads.FollowKey(id, lid),
+			ReadKeyLogID: keyLog,
+		},
+		ThreadID: &pb.ProtoThreadID{ID: id},
+		LogID:    &pb.ProtoPeerID{ID: lid},
+		Record:   pbrec,
+	}
+
+	// Push to each address
+	wg := sync.WaitGroup{}
+	for _, addr := range addrs {
+		wg.Add(1)
+		go func(addr ma.Multiaddr) {
+			p, err := addr.ValueForProtocol(ma.P_P2P)
+			if err != nil {
+				log.Error(err)
+				return
+			}
+			pid, err := peer.IDB58Decode(p)
+			if err != nil {
+				log.Error(err)
+				return
+			}
+
+			log.Debugf("pushing record to %s...", p)
+
+			cctx, _ := context.WithTimeout(ctx, reqTimeout)
+			conn, err := s.dial(cctx, pid, grpc.WithInsecure())
+			if err != nil {
+				log.Error(err)
+				return
+			}
+			client := pb.NewThreadsClient(conn)
+			reply, err := client.Push(cctx, req)
+			if err != nil {
+				log.Error(err)
+				return
+			}
+
+			log.Debugf("reply from %s: %t", p, reply.Ok)
+
+			wg.Done()
+		}(addr)
+	}
+
+	// Finally, publish to the thread's topic
+	err = s.publish(id, req)
+	if err != nil {
+		log.Error(err)
+	}
+
+	wg.Wait()
+	return nil
+}
+
+type records struct {
+	sync.RWMutex
+	m map[cid.Cid]thread.Record
+	s []thread.Record
+}
+
+func newRecords() *records {
+	return &records{
+		m: make(map[cid.Cid]thread.Record),
+		s: make([]thread.Record, 0),
+	}
+}
+
+func (r *records) List() []thread.Record {
+	r.RLock()
+	defer r.RUnlock()
+	return r.s
+}
+
+func (r *records) Store(key cid.Cid, value thread.Record) {
+	r.Lock()
+	if _, ok := r.m[key]; ok {
+		return
+	}
+	r.m[key] = value
+	r.s = append(r.s, value)
+	r.Unlock()
+}
+
+// pull records from log addresses.
+func (s *service) pull(ctx context.Context, id thread.ID, lid peer.ID, offset cid.Cid, settings *tserv.PullSettings) ([]thread.Record, error) {
```

Comment:
This is the client-side pull method. Internally, we need to gather all the addresses known for the given log.